Sources/Certificate Pinning/TrustValidator.swift

```diff
@@ -0,0 +1,102 @@
+    public func canHandle(challenge: AuthenticationChallenge) -> Bool {
+        return challenge.isServerTrustAuthentication && challenge.serverTrust != nil
+    }
+    
+    /// Allows the `TrustValidator` the chance to validate a given `URLAuthenticationChallenge` against it's local certificate.
+    ///
+    /// - Parameters:
+    ///   - challenge: The `URLAuthenticationChallenge` presented to the `URLSession` object with which this validator is associated.
+    ///   - handler: The handler to be called when the challenge is a 'server trust' authentication challenge. For all other types of authentication challenge, this handler will NOT be called.
+    public func handle(challenge: AuthenticationChallenge, handler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
+        guard canHandle(challenge: challenge), let serverTrust = challenge.serverTrust else {
+            return //The challenge was not a server trust evaluation, and so left unhandled
+        }
+        
+        switch evaluate(serverTrust, forHost: challenge.host) {
+        case .allow(let credential): handler(.useCredential, credential)
+        case .block: handler(configuration.authenticationDispositionForFailedValidation(forHost: challenge.host), nil)
+        case .notPinned: handler(.performDefaultHandling, nil)
+        }
+    }
+    
+    func evaluate(_ trust: SecTrust, forHost host: String, date: Date = Date()) -> Decision {
+        guard let domainConfig = configuration.domainConfiguration(forHost: host), domainConfig.shouldValidateCertificate(forHost: host, at: date) else {
+            return .notPinned //We are either not able to retrieve the certificate from the trust or we are not configured to pin this domain
+        }
+        
+        //Set an SSL policy and evaluate the trust
+        let policies = NSArray(array: [SecPolicyCreateSSL(true, nil)])
+        SecTrustSetPolicies(trust, policies)
+        
+        guard trust.isValid else { return .block }
+        
+        //If the server trust evaluation is successful, walk the certificate chain
+        let certificateCount = SecTrustGetCertificateCount(trust)
+        for certIndex in 0..<certificateCount {
+            guard let certificate = SecTrustGetCertificateAtIndex(trust, certIndex) else { continue }
+            
+            if domainConfig.validate(against: certificate) {
+                //Found a pinned certificate, allow the connection
+                return .allow(URLCredential(trust: trust))
+            }
+        }
+        
+        return .block
+    }
+}
+
+// MARK: - SecTrust Utility
+
+extension SecTrust {
+    
+    /// Evaluates `self` and returns `true` if the evaluation succeeds with a value of `.unspecified` or `.proceed`.
+    var isValid: Bool {
+        var result = SecTrustResultType.invalid
+        let status = SecTrustEvaluate(self, &result)
```

Comment:
Potential [code duplication](https://github.com/BottleRocketStudios/iOS-Hyperspace/pull/84/files#diff-ec0b80b5158fa4e1235d54e3968fccb4R112) with the `CertificateHasher`? See comment above about the result of `SecTrustEvaulate(...)` not being used.